Implementation of Cutout layer and RandomErasing layer

chjort opened the pull request:
Implementation of `CutOut` layer and `RandomErasing` layer which closes issues #27 and #39.

Both of these layers still take a `labels` argument in their call method, even though they don't modify the labels.
I was unsure if all augmentation layers should expect a `labels` argument for the sake of consistency, so I let it remain.

For future reference, all layers making use of the `tf.map_fn` with `_fill_rectangle` should probably be refactored to share functionality.

Here are samples:

`RandomCutout(height_factor=0.2, width_factor=0.2, fill_mode="constant", fill_value=0.0)`
![random_cutout_05_05_s](https://user-images.githubusercontent.com/9061122/151267427-9f569019-941e-43ee-b332-7ccc24af40bb.png)

`RandomCutout(height_factor=28, width_factor=90, fill_mode="constant", fill_value=0.0)`
![random_cutout_28_90_s](https://user-images.githubusercontent.com/9061122/151267565-c7b68c62-2378-4769-8653-a32271776402.png)

`RandomCutout(height_factor=(0.3, 0.9), width_factor=64, fill_mode="gaussian_noise")`
![random_cutout_tup_64_s](https://user-images.githubusercontent.com/9061122/151267645-0964531d-b882-4193-b94e-489f1893edad.png)

`RandomCutout(height_factor=(0, 224), width_factor=(0, 224), fill_mode="gaussian_noise")`
![random_cutout_full_s](https://user-images.githubusercontent.com/9061122/151272177-f1a6226d-6831-449e-b07f-0e3714b51894.png)

bhack commented on keras_cv/layers/preprocessing/erasing.py:
```diff
@@ -0,0 +1,226 @@
+            args.append(patch_value)
+
+        images = tf.map_fn(
+            lambda x: _fill_rectangle(*x),
```
How much slow Is this working on single images?
Do we need something working on batches like [other image ops](https://www.tensorflow.org/api_docs/python/tf/image/extract_patches)?

chjort replied:
I have not made any performance tests on using `tf.map_fn`, but we should definitely expect it to be slower compared to a proper vectorized/batched implementation.

Right now, the `CutMix` and `MixUp` layers use `tf.map_fn`, so I used for these layers as well to keep implementations similar.

We have working implementations and tests for these layers now, so we can definitely work towards a vectorized implementation with confidence.

LukeWood commented on keras_cv/layers/preprocessing/erasing.py:
```diff
@@ -0,0 +1,226 @@
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+def _fill_rectangle(
```
Let's refactor this to be in a shared `util.py` file and reuse it across `CutMix` and this.

chjort replied:
I have made a `keras_cv.utils.fill_utils` module containing a `fill_rectangle` function for now. This is not shared between `CutMix`, `CutOut`, and `RandomErasing`.

The decision to make a `keras_cv.utils` package was to mimic the utils structure in `keras`.

LukeWood commented on examples/layers/preprocessing/cut_out_demo.py:
```diff
@@ -0,0 +1,47 @@
+"""cut_out_demo.py shows how to use the CutOut preprocessing layer.
```
Please save some sample images generated by this script and link them in the PR.

LukeWood commented on examples/layers/preprocessing/random_erasing_demo.py:
```diff
@@ -0,0 +1,47 @@
+"""random_erasing_demo.py shows how to use the RandomErasing preprocessing layer.
```
Same here, include some images in the PR description.

thanks!

LukeWood commented on keras_cv/layers/preprocessing/erasing.py:
```diff
@@ -0,0 +1,226 @@
+            patches with gaussian noise. Defaults to 0.0.
+    """
+
+    def __init__(self, rate, patch_value=None, seed=None, **kwargs):
```
Default the value in the constructor

LukeWood commented on keras_cv/layers/preprocessing/erasing.py:
```diff
@@ -0,0 +1,226 @@
+
+    def get_config(self):
+        config = {"length": self.length}
+        base_config = super(CutOut, self).get_config()
```
super()

LukeWood commented on keras_cv/layers/preprocessing/erasing.py:
```diff
@@ -0,0 +1,226 @@
+        rate: Float between 0 and 1.  The fraction of samples to augment.
+        scale: Tuple of float. Area scale range (min, max) of erasing patch.
+        ratio: Tuple of float. Aspect ratio range (min, max) of erasing patch.
+        patch_value: Float. The value to fill in the patches. If None, will
```
Defaults to `None`.  Let's also give this a value instead of `None`.  Maybe `noise` or `random`.

chjort replied:
I have decided on `"gaussian"`.

fchollet commented on keras_cv/layers/preprocessing/erasing.py:
```diff
@@ -0,0 +1,196 @@
+from keras_cv.utils import fill_utils
+
+
+class BaseErasing(layers.Layer, abc.ABC):
```
We should not depend on ABC for base layers (see other base layers in the Keras codebase for examples). `@abc.abstractmethod` decorators should be replaced with `NotImplementedError`.

fchollet replied:
In this specific case I also recommend *not* having this base layer at all, instead having independent augmentation layers. Most of the shared code can be moved to stateless utility functions (e.g. `_erase` which would take a patch size computation function). Some of the shared code should be duplicated across layers to improve code readability: in particular all layers should have their own `call()` method. Explicit is better than implicit.

In terms of file structure, I recommend putting each layer of the 2 layers in its own file, named after the layer name.

chjort replied:
I see. The layers are now independent, and each have their own file.

fchollet commented on keras_cv/layers/preprocessing/erasing.py:
```diff
@@ -0,0 +1,196 @@
+
+    Args:
+        rate: Float between 0 and 1.  The fraction of samples to augment.
+        patch_value: Float or string. The value to fill in the patches. If "gaussian", will
```
This is similar to the `fill_mode` argument in other augmentation layers. Let's call it that. Initial values should be `"gaussian_noise"` or `"constant"`. Add a `fill_value=0.0` argument for the constant mode.

fchollet commented on keras_cv/layers/preprocessing/erasing.py:
```diff
@@ -0,0 +1,196 @@
+        self.patch_value = patch_value
+        self.seed = seed
+
+    def call(self, images, labels):
```
All augmentation layers should have a `training=True` argument. When False (inference mode) no augmentation is applied.

fchollet commented on keras_cv/layers/preprocessing/erasing_test.py:
```diff
@@ -0,0 +1,163 @@
+import tensorflow as tf
+
+from keras_cv.layers.preprocessing.erasing import CutOut, RandomErasing
```
One import per line, also don't import specific symbols, import modules (like `erasing`).

fchollet reviewed (commented):
Thanks for the PR! 👍 

Is this based on https://github.com/tensorflow/similarity/blob/master/tensorflow_similarity/augmenters/img_augments.py ?

LukeWood commented on keras_cv/layers/preprocessing/erasing.py:
````diff
@@ -0,0 +1,196 @@
+    ```
+    """
+
+    def __init__(self, rate, length, patch_value=0.0, seed=None, **kwargs):
````
Maybe we should actually make length a float from 0-1 that represents the relative width/height of the image?

So for an image of (150, 150) a length of 0.5 is equivalent to 75.

LukeWood replied:
(this applies to both classes)

Also open to discussion here, @fchollet may have some thoughts.  One benefit of using a float in range [0, 1] is that it can be copy pasted into new code bases in a meaningful way.  As a reader the number 0.4 is also very meaningful to me: 40% of the image.  40 could mean anything from covering the entire image to covering 0.01% of a high res image.

LukeWood replied:
maybe `size`, or `width` would be good as well?

chjort replied:
I can follow your logic. A scale between [0, 1] would also make `CutOut` follow the same pattern as `RandomErasing`, increasing consistency.

However, the [CutOut paper](https://arxiv.org/pdf/1708.04552.pdf) only denotes sizes in terms of absolute values, while the [RandomErasing paper](https://arxiv.org/pdf/1708.04896.pdf) actually denotes sizes in relative values. If a user wants to draw references to the papers, it might be confusing for users if the size formats differ.

Perhaps we should weigh consistency higher?

LukeWood replied:
I'd prefer to weight consistency higher here.  We may also want to use max_aspect and min_aspect.

I do quite like the name aspect.

chjort replied:
Actually, CutOut is always a square patch per the official paper. It will be confusing to specify a square patch using a scalar between [0, 1] as relative width/height of input image, if input image is non-square.

LukeWood replied:
Ah very interesting.  I don't personally see a reason we need to match the implementation in that regard - though perhaps @fchollet may?

chjort replied:
> Ah very interesting. I don't personally see a reason we need to match the implementation in that regard - though perhaps fchollet may?

@LukeWood Let me know if I should make any changes to how the patch size is specified, and what change you would prefer. Unless we are waiting for fchollet's opinion?

LukeWood replied:
Yeah, let's wait for @fchollet 's opinion here.  He also mentioned that we may want to merge the layers and rely just on the fill arg.  He said he would do a pass on this PR in a bit.

LukeWood commented on examples/layers/preprocessing/cut_out_demo.py:
```diff
@@ -0,0 +1,47 @@
+"""
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from keras_cv.layers.preprocessing import erasing
```
Recently I pushed a commit importing Preprocessing layers to the __init__.py file for `keras_cv/layers/preprocessing/__init__.py`.  Let's also import these layers there, and then we can do:

```
from keras_cv.layers import preprocessing
preprocessing.CutOut
```

LukeWood commented:
@chjort almost there, thanks again for the PR - this is great stuff.

chjort commented:
@fchollet
> Thanks for the PR! +1
> 
> Is this based on https://github.com/tensorflow/similarity/blob/master/tensorflow_similarity/augmenters/img_augments.py ?

Yes, the layers are based on the above, especially making use of the `_fill_rectangle` function.

I have made a vectorized/batched implementation of `fill_rectangle`, but i still need to make some performance tests. I plan to include it in a separate PR after this one is merged.

chjort commented:
@fchollet, @LukeWood 

What are your opinions on the `labels` argument in `call` for layers that do not modify lables? I kept the `labels` argument in these layers to keep them consistent with `MixUp` and `CutMix`. 

Do we have any design decision regarding labels in these preprocessing layers going forward? Ultimately we will need it to support bounding boxes or segmentation masks.

LukeWood commented:
> @fchollet, @LukeWood
> 
> What are your opinions on the `labels` argument in `call` for layers that do not modify lables? I kept the `labels` argument in these layers to keep them consistent with `MixUp` and `CutMix`.
> 
> Do we have any design decision regarding labels in these preprocessing layers going forward? Ultimately we will need it to support bounding boxes or segmentation masks.

ah - yes please drop the labels argument!  MixUp and CutMix are a special case.

LukeWood commented:
> I have made a vectorized/batched implementation of fill_rectangle, but i still need to make some performance tests. I plan to include it in a separate PR after this one is merged.

That's great!  Thank you!

bhack commented:
@LukeWood Does `fill_rectange` API need to handle a batch of image with a batch of rectangles?

LukeWood commented:
> @LukeWood Does `fill_rectange` API need to handle a batch of images with a batch of rectangles?

Ideally, yeah we should be able to batch calls to `fill_rectangle`.  Do we have a batched implementation of this?

chjort commented:
I have removed the `labels` argument from `CutOut` and `RandomErasing`. I also added the missing `training=True` argument to `CutMix` and `MixUp`.

I pushed these changes from one of my other machines, which apparently uses one of my other email adresses, therefore the CLA check fails. I hope it is alright.

LukeWood commented:
@chjort is there any real difference between these two layers other than fill mode & the width/ratio arguments at this point?  Perhaps these should actually be merged into a single layer.

chjort commented:
There really is no difference between these layers except for how size and shape of patches are specified. Fill mode is the same between the two layers, just different default values. It is definitly possible to unite these into a single layer e.g. `FillRectangle` while supporting both types of patches.

But, it really depends on what the goal is regarding the building blocks that KerasCV implements. Do we make building blocks that implicitly supports SOTA and specialized CV techniques, or do we make explicit building blocks of known SOTA techniques? I.e. do we want users to make the connection that `FillRectangle` layer should be used to achieve [CutOut](https://arxiv.org/pdf/1708.04552.pdf) and [RandomErasing](https://arxiv.org/pdf/1708.04896.pdf), and possibly any future similar type of augmentation? Or should these known techniques be explicitly implemented individually?

You could also argue that CutMix and MixUp are very similar, and only differs in mixing area, and mixing ratio. Hence, the above thoughts could also be applied to these two layers.

LukeWood commented:
> @LukeWood
> There really is no difference between these layers except for how size and shape of patches are specified. Fill mode is the same between the two layers, just different default values. It is definitly possible to unite these into a single layer e.g. FillRectangle while supporting both types of patches.
> 
> But, it really depends on what the goal is regarding the building blocks that KerasCV implements. Do we make building blocks that implicitly supports SOTA and specialized CV techniques, or do we make explicit building blocks of known CV techniques? I.e. do we want users to make the connection that FillRectangle layer can be used to achieve CutOut and RandomErasing, and possibly any future similar type of augmentation, or should these known techniques be explicitly implemented individually?
> 
> You could also argue that CutMix and MixUp are very similar, and only differs in mixing area, and mixing ratio. Hence, the above reflections could also be relevant to these two layers.
> 

Yeah, this is a pretty nuanced decision.  I personally would propose we have a single layer with a fill mode argument, that takes aspect, which can be a static float [0, 1] or a tuple of floats (aspect_weight, aspect_height).  In my opinion this minimizes code we have to maintain, allows us to support every possible use case, and is a clear user API.  It's not clear to me that the square part of CutOut is actually important to its performance - I'd actually anticipate it to perform worse given how it modifies the training distribution.  

Let's delegate to @fchollet on this decision though as it is pretty nuanced.

LukeWood commented:
Thank you again for your patience and contribution - this is really helpful and we really appreciate the code!

fchollet commented:
> I personally would propose we have a single layer with a fill mode argument, that takes aspect, which can be a static float [0, 1] or a tuple of floats (aspect_weight, aspect_height). In my opinion this minimizes code we have to maintain, allows us to support every possible use case, and is a clear user API. It's not clear to me that the square part of CutOut is actually important to its performance - I'd actually anticipate it to perform worse given how it modifies the training distribution.

I agree -- what matters is to make available building blocks you can use to build state-of-the-art pipelines. Whether they match the exact specifications listed in the first paper that described them is irrelevant.

"Randomly cutting out rectangles and filling them with something" is the general category we should be covering, and then we should provide an optimal default value for "what to fill them for". Both constant values and noise seem suboptimal by the way. Hopefully in the future we can provide a default that fills in the patch with content generated from the rest of the image.

fchollet commented on keras_cv/layers/preprocessing/cut_out.py:
```diff
@@ -0,0 +1,148 @@
+        name=None,
+        **kwargs
+    ):
+        super().__init__(name=name, **kwargs)
```
You don't need to list `name` in the signature explicitly, you can just leave it as part of `kwargs`

LukeWood commented:
> > I personally would propose we have a single layer with a fill mode argument, that takes aspect, which can be a static float [0, 1] or a tuple of floats (aspect_weight, aspect_height). In my opinion this minimizes code we have to maintain, allows us to support every possible use case, and is a clear user API. It's not clear to me that the square part of CutOut is actually important to its performance - I'd actually anticipate it to perform worse given how it modifies the training distribution.
> 
> I agree -- what matters is to make available building blocks you can use to build state-of-the-art pipelines. Whether they match the exact specifications listed in the first paper that described them is irrelevant.
> 
> "Randomly cutting out rectangles and filling them with something" is the general category we should be covering, and then we should provide an optimal default value for "what to fill them for". Both constant values and noise seem suboptimal by the way. Hopefully in the future we can provide a default that fills in the patch with content generated from the rest of the image.

Thanks for the reply.  Yeah, sorry for any confusion @chjort let's move to a single layer that follows the specification above.  For now, don't worry about the last point of the fill mode.  There is a specific contributor that will likely take a look at that.

bhack commented:
> Both constant values and noise seem suboptimal by the way. Hopefully in the future we can provide a default that fills in the patch with content generated from the rest of the image.

Could it be something like [tf.pad](https://www.tensorflow.org/api_docs/python/tf/pad) with its `mode`? There we have an extra `constant_values=0` arg that it is used only for a specific mode.

LukeWood commented:
> Could it be something like tf.pad with its mode? There we have an extra constant_values=0 arg that it is used only for a specific mode.

For now this is fine, there is a specific contributor that is interested in contributing a `mirror` fill mode as well.

fchollet reviewed (commented):
We can just keep a single layer and name is `RandomCutout`. The `Random` prefix is for consistency with other random augmentation layers.

I'd recommend:

`RandomCutout(height_factor, width_factor, fill_mode='constant', fill_value=0.0, rate=1.0)`

Then the cutout area is a random rectangle where the rectangle side is drawn uniformly from `[0, img_side * factor]` for each side (when the factor is a float between 0 and 1). When the factor is an integer it is interpreted as the max value in a pixel range.

bhack commented:
> there is a specific contributor that is interested in contributing a mirror fill mode as well.
Do you mean an internal contributor or an external contributor from a public ticket/comment on github?

LukeWood commented:
> Do you mean an internal contributor or an external contributor from a public ticket/comment on github?

internal

bhack commented:
Linting is failing. I hope that we could adopt some form of pre-commit hooks.

chjort commented:
Great discussion! I have combined it into a single layer named `RandomCutout`.

The layer now takes a `height_factor` argument and `width_factor` argument. These arguments supports specifying a single float as fraction of input image, and single integer as absolute pixel size of side. Each argument also accept a tuple of size 2 representing an upper and lower bound range. Read docstring for more details.

If you think complexity becomes too high by allowing tuple ranges in the arguments, then let me know and I will change it to only single floats/ints.

LukeWood commented on keras_cv/layers/preprocessing/cut_mix.py:
```diff
@@ -53,7 +56,7 @@ def _sample_from_beta(alpha, beta, shape):
         return sample_alpha / (sample_alpha + sample_beta)
 
-    def call(self, images, labels):
+    def call(self, images, labels, training=True):
```
Thanks for adding this by the way

LukeWood commented:
I like it - thanks for the great contribution @chjort

LukeWood commented on keras_cv/layers/preprocessing/random_cutout.py:
```diff
@@ -0,0 +1,248 @@
+            self.height_lower = height_factor[0]
+            self.height_upper = height_factor[1]
+        else:
+            self.height_lower = height_factor
```
I'd like to actually default this to 0

Same with width.  I can push this change myself though.

chjort commented:
@LukeWood  You are welcome!
